hathor/dag_builder/tokenizer.py

```diff
@@ -0,0 +1,170 @@
+    a.attr = value              # set value of attribute attr to a
+
+Special attributes:
+    a.out[i] = 100 HTR          # set that the i-th output of a holds 100 HTR
+    a.out[i] = 100 TOKEN        # set that the i-th output of a holds 100 TOKEN where TOKEN is a custom token
+
+
+Example:
+
+    blockchain genesis a[0..300]
+    blockchain a300 b[0..20]
+    blockchain b4 c[0..10]
+
+    # reward lock
+    a300 < dummy
+
+    b11 --> tx1
+    b11 --> tx2
+
+    b14 --> tx1
+    b14 --> tx3
+
+    c3 --> tx1
+    c3 --> tx2
+
+    tx1 <-- tx2 <-- tx3
+
+    tx3 --> tx5 --> tx6
+
+    tx1.out[0] <<< tx2 tx3
+    tx1.out[0] <<< tx4
+
+    a0.out[0] <<< tx1
+
+    tx1.out[0] = 100 HTR [wallet1]
+    tx1.out[1] = 50 TK1  [wallet2]
+    tx2.out[0] = 75 USDC [wallet1]
+
+    USDC.out[0] = 100000 HTR
+
+    b5 < c0 < c10 < b20
+    b6 < tx3
+    b16 < tx4
+"""
+
+
+class TokenType(Enum):
+    BLOCKCHAIN = auto()
+    ATTRIBUTE = auto()
+    PARENT = auto()
+    SPEND = auto()
+    OUTPUT = auto()
+    ORDER_BEFORE = auto()
+
+
+Token = tuple[TokenType, tuple[Any, ...]]
+
+
+def collect_pairs(parts: list[str], expected_sep: str) -> Iterator[tuple[str, str]]:
+    n = len(parts)
+    if n < 3:
+        raise SyntaxError
+    if n % 2 == 0:
+        raise SyntaxError
+
+    k = (n - 1) // 2
+    for i in range(k):
+        first = parts[2 * i]
+        sep = parts[2 * i + 1]
+        second = parts[2 * i + 2]
+        if parts[2 * i + 1] != expected_sep:
+            raise SyntaxError(f'inconsistent separator; got {sep} but expecting {expected_sep}')
+        yield (first, second)
+
+
+def tokenize(content: str) -> Iterator[Token]:
+    """Parse content and generate tokens.
+    """
+    blockchain_re = re.compile(r'^([a-zA-Z][a-zA-Z0-9-_]*)\[([0-9]+)..([0-9]+)\]$')
+    first_parent: str | None
+    for line in content.split('\n'):
+        line, _, _ = line.partition('#')
+        line = line.strip()
+        if not line:
+            continue
+
+        # split() trims on both sides and remove empty parts
+        parts = line.split()
+
+        if parts[0] == 'blockchain':
+            if len(parts) != 3:
+                raise SyntaxError
+            first_parent = parts[1]
+            if first_parent == 'genesis':
+                first_parent = None
+            match = blockchain_re.match(parts[2])
+            if not match:
+                raise SyntaxError(f'invalid blockchain format: {line}')
+            name, begin, end = match.groups()
+            yield (TokenType.BLOCKCHAIN, (name, first_parent, int(begin), int(end)))
+
+        elif parts[1] == '=':
+            name, key = parts[0].split('.', 1)
+            if key.startswith('out[') and key[-1] == ']':
+                index = int(key[4:-1])
+                amount = int(parts[2])
+                token = parts[3]
+                attrs = parts[4:]
+                yield (TokenType.OUTPUT, (name, index, amount, token, attrs))
```

Comment:
What are `attrs` in this case? I think it's not mentioned in the example.

Reply:
We can use them for timestamp and other stuff. But the first use I'll give is on nanos in another PR.